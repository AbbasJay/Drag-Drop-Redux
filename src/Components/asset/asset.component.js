import React, { Component } from "react";
import { placeItem, removeItem, addDraggableItem } from "../../Actions/action";
import { connect } from "react-redux";
import "./asset.styles.css";

export class Asset extends Component {
  dragStartHandler = () => {
    this.props.addDraggableItem(this.props.index);
  };
  handleClick = () => {
    const { index, placeItem, isSelectedItem } = this.props;
    if (!isSelectedItem) {
      placeItem(index);
    }
  };
  render() {
    const { item, index, isSelectedItem, removeItem } = this.props;
    return (
      item && (
        <li
          className={"asset type-" + item.type}
          onClick={this.handleClick}
          draggable
          onDragStart={() => this.dragStartHandler()}
        >
          <h2>{item.name}</h2>
          <p>
            <span className="info">
              {item.duration + " "}
              {item.durationType && item.durationType}
            </span>
            <span className="info">{item.type}</span>
          </p>
          {isSelectedItem && (
            <button
              onClick={() => {
                removeItem(index);
              }}
              className="remove-asset"
            >
              x
            </button>
          )}
        </li>
      )
    );
  }
}

const mapActionsToProps = dispatch => {
  return {
    placeItem: index => dispatch(placeItem(index)),
    removeItem: index => dispatch(removeItem(index)),
    addDraggableItem: index => dispatch(addDraggableItem(index))
  };
};

export default connect(
  null,
  mapActionsToProps
)(Asset);
