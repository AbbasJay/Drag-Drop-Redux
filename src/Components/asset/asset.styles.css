li.asset p {
  margin: 0;
}

li.type-article {
  background: #131540;
}

li.type-document {
  background: #d4ae00;
}

li.type-video {
  background: #f14b5a;
}

li.type-online-course {
  background: #4bd13f;
}
