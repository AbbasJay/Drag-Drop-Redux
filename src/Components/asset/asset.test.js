import React from "react";
import { Asset } from "./asset.component";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

const removeItemFn = jest.fn();
const placeItemFn = jest.fn();
const addDraggableItemFn = jest.fn();
const assetProps = {
  placeItem: placeItemFn,
  removeItem: removeItemFn,
  addDraggableItem: addDraggableItemFn,
  isSelectedItem: true,
  item: {},
  index: 0
};

Enzyme.configure({ adapter: new Adapter() });

describe("asset component", () => {
  it("renders", () => {
    const div = shallow(<Asset {...assetProps} />);
    expect(div).toMatchSnapshot();
  });

  it("renders x when isSelected is true", () => {
    const wrapper = shallow(<Asset {...assetProps} />);
    expect(wrapper.find("button")).toHaveLength(1);
  });

  it("when x is clicked it should call removeItem()", () => {
    const wrapper = shallow(<Asset {...assetProps} />);
    const xButton = wrapper.find("button");
    xButton.simulate("click");
    expect(removeItemFn).toHaveBeenCalledWith(0);
  });

  it("does not render x when isSelected is false", () => {
    const renderedAsset = shallow(<Asset {...assetProps} />);
    renderedAsset.setProps({ isSelectedItem: false });
    expect(renderedAsset.find("button")).toHaveLength(0);
  });

  it("when element is clicked expect placeItem() to be called", () => {
    const renderedAsset = shallow(<Asset {...assetProps} />);
    renderedAsset.setProps({ isSelectedItem: false });
    renderedAsset.simulate("click");
    expect(placeItemFn).toHaveBeenCalledWith(0);
  });

  it("when element is dragged addDraggableItem() to be called", () => {
    const renderedAsset = shallow(<Asset {...assetProps} />);
    renderedAsset.setProps({ isSelectedItem: false });
    renderedAsset.simulate("dragStart");
    expect(addDraggableItemFn).toHaveBeenCalledWith(0);
  });
});
