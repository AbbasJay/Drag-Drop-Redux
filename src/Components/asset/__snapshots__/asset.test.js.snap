// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asset component renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Asset
    addDraggableItem={[MockFunction]}
    index={0}
    isSelectedItem={true}
    item={Object {}}
    placeItem={[MockFunction]}
    removeItem={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2 />,
        <p>
          <span
            className="info"
          >
            undefined 
          </span>
          <span
            className="info"
          />
        </p>,
        <button
          className="remove-asset"
          onClick={[Function]}
        >
          x
        </button>,
      ],
      "className": "asset type-undefined",
      "draggable": true,
      "onClick": [Function],
      "onDragStart": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="info"
            >
              undefined 
            </span>,
            <span
              className="info"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "undefined ",
                undefined,
              ],
              "className": "info",
            },
            "ref": null,
            "rendered": Array [
              "undefined ",
              undefined,
            ],
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
              "className": "info",
            },
            "ref": null,
            "rendered": null,
            "type": "span",
          },
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "x",
          "className": "remove-asset",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "x",
        "type": "button",
      },
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2 />,
          <p>
            <span
              className="info"
            >
              undefined 
            </span>
            <span
              className="info"
            />
          </p>,
          <button
            className="remove-asset"
            onClick={[Function]}
          >
            x
          </button>,
        ],
        "className": "asset type-undefined",
        "draggable": true,
        "onClick": [Function],
        "onDragStart": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="info"
              >
                undefined 
              </span>,
              <span
                className="info"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "undefined ",
                  undefined,
                ],
                "className": "info",
              },
              "ref": null,
              "rendered": Array [
                "undefined ",
                undefined,
              ],
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
                "className": "info",
              },
              "ref": null,
              "rendered": null,
              "type": "span",
            },
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "x",
            "className": "remove-asset",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "x",
          "type": "button",
        },
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
